R/type2.R

```diff
@@ -409,3 +422,41 @@ vec_is_subtype <- function(x, super, ..., x_arg = "", super_arg = "") {
 vec_implements_ptype2 <- function(x) {
   .Call(vctrs_implements_ptype2, x)
 }
+
+# Example usage of ptype2 restart. As there is no cast restart (yet),
+# it requires `x` and `y` to be convertible to the restarted ptype.
+with_ordered_restart <- function(expr) {
+  withCallingHandlers(
+    expr,
+    vctrs_error_ptype2 = function(cnd) {
+      x <- cnd[["x"]]
+      y <- cnd[["y"]]
+
+      restart <- FALSE
+
+      if (is.ordered(x)) {
+        restart <- TRUE
+        x <- as.character(x)
```

Comment:
I guess the way this is set up allows for `<ordered>` `x` values to gain access to _all_ of the `<character>` ptype2 methods.

What I'm trying to say is that even though we designed this for the case of: `<ordered> + <factor> = error` is retried as `<character> + <factor> = <character>`, it is really set up so that `<ordered> + <any>` can be retried as `<character> + <any>`.

It would be up to the author of this handler to check for `is.ordered(x) && is.factor(y)` if they want it to be more specific

Reply:
`ordered + any` is now retried as `factor + any` to preserve factorness if possible. I think simplicity is more important than conservatiness for this handler, so it doesn't try to prevent inheriting factor methods.